processing/src/main/java/io/druid/query/topn/TopNQueryQueryToolChest.java

```diff
@@ -306,27 +303,15 @@ public String apply(@Nullable AggregatorFactory input)
       @Override
       public byte[] computeCacheKey(TopNQuery query)
       {
-        final byte[] dimensionSpecBytes = query.getDimensionSpec().getCacheKey();
-        final byte[] metricSpecBytes = query.getTopNMetricSpec().getCacheKey();
-
-        final DimFilter dimFilter = query.getDimensionsFilter();
-        final byte[] filterBytes = dimFilter == null ? new byte[]{} : dimFilter.getCacheKey();
-        final byte[] aggregatorBytes = QueryCacheHelper.computeAggregatorBytes(query.getAggregatorSpecs());
-        final byte[] granularityBytes = query.getGranularity().cacheKey();
-
-        return ByteBuffer
-            .allocate(
-                1 + dimensionSpecBytes.length + metricSpecBytes.length + 4 +
-                granularityBytes.length + filterBytes.length + aggregatorBytes.length
-            )
-            .put(TOPN_QUERY)
-            .put(dimensionSpecBytes)
-            .put(metricSpecBytes)
-            .put(Ints.toByteArray(query.getThreshold()))
-            .put(granularityBytes)
-            .put(filterBytes)
-            .put(aggregatorBytes)
-            .array();
+        return new CacheKeyBuilder(TOPN_QUERY, CacheKeyBuilder.EMPTY_BYTES)
+            .appendCacheable(query.getDimensionSpec())
+            .appendCacheable(query.getTopNMetricSpec())
+            .appendInt(query.getThreshold())
+            .appendCacheable(query.getGranularity())
+            .appendCacheable(query.getDimensionsFilter())
+            .appendCacheableList(query.getAggregatorSpecs())
+            .appendCacheableList(query.getPostAggregatorSpecs())
```

Comment:
Until now, just changing the post-aggregators on a topN (maybe adding new arithmetic, something like that) would still allow the cache to be used. This is a bug if we're sorting by the post-aggregator but is fine otherwise.

What do you think about changing the cache key generation so it only includes post-aggs we're sorting on (and any transitive dependencies based on `getDependentFields`)?